core/src/main/java/apoc/trigger/TriggerHandlerRead.java

```diff
@@ -0,0 +1,228 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static apoc.ApocConfig.apocConfig;
+import static apoc.trigger.TriggerHandlerWrite.getTriggerNodes;
+import static apoc.trigger.TriggerHandlerWrite.withSystemDb;
+
+
+public class TriggerHandlerRead extends LifecycleAdapter implements TransactionEventListener<Void> {
+
+    public static final int TRIGGER_RELOAD_DEFAULT_VALUE = 3000;
+    public static final String TRIGGER_RELOAD = "apoc.trigger.reload";
+
+    private enum Phase {before, after, rollback, afterAsync}
+    
+    private final ConcurrentHashMap<String, Map<String,Object>> activeTriggers = new ConcurrentHashMap<>();
+
+    private final Log log;
+    private final GraphDatabaseService db;
+    private final DatabaseManagementService databaseManagementService;
+    private final Pools pools;
+    
+    private final JobScheduler jobScheduler;
+
+    private long lastUpdate;
+
+    private JobHandle restoreTriggerHandler;
+
+    private final AtomicBoolean registeredWithKernel = new AtomicBoolean(false);
+
+    public TriggerHandlerRead(GraphDatabaseService db, DatabaseManagementService databaseManagementService,
+                              Log log, Pools pools, JobScheduler jobScheduler) {
+        this.db = db;
+        this.databaseManagementService = databaseManagementService;
+        this.log = log;
+        this.pools = pools;
+        this.jobScheduler = jobScheduler;
+    }
+
+
+    public void reconcileKernelRegistration() {
+        final boolean hasTriggers = withSystemDb(tx -> getTriggerNodes(db.databaseName(), tx)
+                .hasNext());
+        reconcileKernelRegistration(hasTriggers);
+    }
+
+    /**
+     * There is substantial memory overhead to the kernel event system, so if a user has enabled apoc triggers in
+     * config, but there are no triggers set up, unregister to let the kernel bypass the event handling system.
+     *
+     * For most deployments this isn't an issue, since you can turn the config flag off, but in large fleet deployments
+     * it's nice to have uniform config, and then the memory savings on databases that don't use triggers is good.
+     */
+    public synchronized void reconcileKernelRegistration(boolean hasTriggers) {
+        // Register if there are triggers
+        if (hasTriggers) {
+            // This gets called every time triggers update; only register if we aren't already
+            if(registeredWithKernel.compareAndSet(false, true)) {
+                databaseManagementService.registerTransactionEventListener(db.databaseName(), this);
+            }
+        } else {
+            // This gets called every time triggers update; only unregister if we aren't already
+            if(registeredWithKernel.compareAndSet(true, false)) {
+                databaseManagementService.unregisterTransactionEventListener(db.databaseName(), this);
+            }
+        }
+    }
```

Comment:
I thought I had written a comment about this, but I've been jumping in and out of this review all day and can't find it anymore 😢 

I'm not sure that this whole section would work, or perhaps I'm misunderstanding what it's supposed to do. This code seems to be called whenever we install a new trigger, which means it's being run in a transaction against the "system" database. At that point wouldn't this be registered against the system database which might not even have a given user database (in 5.x at least, in 4.x they always will)?

Shouldn't this logic rather be run in the background jobs happening for each user database for which there are triggers?

Reply:
Oh I see that this code is being called from multiple places... I think I get how it's supposed to work now. The start() method is gonna get called on each DBMS which has APOC installed, and then that's gonna call reconcileKernelRegistration.

What might still not make sense to me is why this method is being called during apoc.trigger.install, apoc.trigger.drop, and apoc.trigger.dropAll when that code is executed in a "system" transaction on a DBMS which in 5.x might not even have a given user database.

Reply:
Ah thanks, I edited my original answer I think, maybe just before you posted your reply. In any case I think we're now on a similar understanding of how this works. That part of the code is fine and great.

What still doesn't make much sense to me is why reconcileKernelRegistration() is being called in the controller class because in the code in the controller class is basically being executed while we're in a transaction for the system database.

Example of unnecessary/harmful invocations of this method:
- [here in apoc.trigger.install](https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/3183/files#diff-28ed54275ec240e9acd877b090ba248e89e4f67e3083f086fef57b2ddbce5c5aR100)
- [here in apoc.trigger.drop](https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/3183/files#diff-28ed54275ec240e9acd877b090ba248e89e4f67e3083f086fef57b2ddbce5c5aR117)
- [here in apoc.trigger.dropAll](https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/3183/files#diff-28ed54275ec240e9acd877b090ba248e89e4f67e3083f086fef57b2ddbce5c5aR133)

If there is still confusion from either of us after this then maybe a call will be the best way to resolve things.

Reply:
I've changed the caching logic. Now new and deprecated procedure use the same `TriggerHandler`, which has a private `reconcilceKernelRegistration()`

Reply:
Beautiful, cheers!